jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CrossOriginHandler.java

```diff
@@ -0,0 +1,423 @@
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("preflight cross-origin request not delivered to child handler {}", request);
+                callback.succeeded();
+                return true;
+            }
+        }
+        return super.handle(request, response, callback);
+    }
+
+    private boolean originMatches(String origins)
+    {
+        if (anyOriginAllowed)
+            return true;
+        if (allowedOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedOriginPatterns);
+    }
+
+    private boolean timingOriginMatches(String origins)
+    {
+        if (anyTimingOriginAllowed)
+            return true;
+        if (allowedTimingOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedTimingOriginPatterns);
+    }
+
+    private boolean originMatches(String origins, Set<Pattern> allowedOriginPatterns)
+    {
+        for (String origin : origins.split(" "))
+        {
+            origin = origin.trim();
+            if (origin.isEmpty())
+                continue;
+            for (Pattern pattern : allowedOriginPatterns)
+            {
+                if (pattern.matcher(origin).matches())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isPreflight(Request request)
+    {
+        return HttpMethod.OPTIONS.is(request.getMethod()) && request.getHeaders().contains(HttpHeader.ACCESS_CONTROL_REQUEST_METHOD);
+    }
+
+    private void handlePreflightResponse(String origin, Response response)
+    {
+        HttpFields.Mutable headers = response.getHeaders();
+        headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
+        if (isAllowCredentials())
+            headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+        Set<String> allowedMethods = getAllowedMethods();
+        if (!allowedMethods.isEmpty())
+            headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_METHODS, String.join(",", allowedMethods));
+        Set<String> allowedHeaders = getAllowedHeaders();
+        if (!allowedHeaders.isEmpty())
+            headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_HEADERS, String.join(",", allowedHeaders));
+        long seconds = getPreflightMaxAge().toSeconds();
+        if (seconds > 0)
+            headers.put(HttpHeader.ACCESS_CONTROL_MAX_AGE, seconds);
+    }
+
+    private void handleSimpleResponse(String origin, Response response)
+    {
+        HttpFields.Mutable headers = response.getHeaders();
+        headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
+        if (isAllowCredentials())
+            headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+        Set<String> exposedHeaders = getExposedHeaders();
+        if (!exposedHeaders.isEmpty())
+            headers.put(HttpHeader.ACCESS_CONTROL_EXPOSE_HEADERS, String.join(",", exposedHeaders));
```

Comment:
ditto pre-encoded

Reply:
This one is rare, I left it out.

Reply:
still should be preencoded.  trivial to do.